### PR TITLE
Redesign the read interface function for AT commands

### PR DESCRIPTION
**Pull request by azeemshatp**
What does this PR do?
- Modify the read_resp() function to handle both reading and checking the expected response in the same function.
- Read buffer is now accessible from outside the library (from the main program) by passing a pointer to the buffer. 
- Removed the function check_resp().
- Modify error handling for few AT commands.

What are the relevant tickets?
NA

Where should the reviewer start?
SIM800-AT.cpp

How should this be manually tested?
Access the library functions from a main program and flash the firmware to the device connected to a SIM800 GSM mode.

Any background context you want to provide? NA
Screenshots (if appropriate)

### CODE REVIEW

File: src/SIM800-AT.cpp [hunk 1]
```diff
@@ -20,329 +20,385 @@
...
+    int actual_ack_num_bytes = 0;
+
+    while (t.read() < time_out) {
+        if (index_buf < size_buf-1) {
             if (gprsSerial.readable()) {
-                rx_buffer[pos] = gprsSerial.getc();
-                pos++;
+                resp_buf[index_buf] = gprsSerial.getc();
+
+                if (0 != strcmp(ack_message, NULL)) { //Decide if an acknowledgement check required
```
Comment (martinjaeger) [inline src/SIM800-AT.cpp]: `strcmp` doesn't make sense here. Please make sure you understand what this function is for.

File: src/SIM800-AT.cpp [hunk 2]
```diff
@@ -20,329 +20,385 @@
...
-        &timeinfo.tm_year, &timeinfo.tm_mon, &timeinfo.tm_mday,
-        &timeinfo.tm_hour, &timeinfo.tm_min, &timeinfo.tm_sec,
-        &timezone);
-
-    if (items == 7) {                   // all information found
-        timeinfo.tm_year += 100;        // struct tm starts counting years from 1900
-        timeinfo.tm_mon -= 1;           // struct tm starts counting months from 0 for January
-        uint32_t timestamp = mktime(&timeinfo); // - timezone * 15 * 60;     // SIM800 provides timezone as multiple of 15 mins but if using NTP, time will always be the specified zone
-        return timestamp;
+    resp_buf[0] = '\0'; // Null char to clear the buffer
```
Comment (martinjaeger) [inline src/SIM800-AT.cpp]: It seems like `read_resp` always responds with a null-terminated buffer anyway now. So this statement is not necessary. (same for all other cases below).

File: src/SIM800-AT.cpp [hunk 3]
```diff
@@ -20,329 +20,385 @@
...
+    read_resp(resp_buf, size_buf, DEFAULT_TIMEOUT, NULL_STR);
+    if (0 != strlen(resp_buf)) {
+        // resp_buf[size_buf-1] = '\0';  // make sure we have a null-terminated string for sscanf
```
Comment (martinjaeger) [inline src/SIM800-AT.cpp]: Delete comment.

File: src/SIM800-AT.cpp (hunk 1 above)
Comment (azeemshatp) [reply src/SIM800-AT.cpp]: Agreed. That was a mistake and corrected now.

**Comment by martinjaeger** [comment]
Is this ready for review again after the latest force-push? If yes, please re-request review (top right corner).

File: src/SIM800-AT.h [hunk 4]
```diff
@@ -44,47 +44,72 @@ class GPRS
...
+     * Valid response   : 0
+     * When used only to read-out the buffer, the function always returns -1.
+     */
+    int read_resp(char *resp_buf, int size_buf, int time_out, const char *ack_message);
+
+    int init(char *resp_buf, int size_buf);
+    int wakeup(char *resp_buf, int size_buf);
+    int check_pin(char *resp_buf, int size_buf);
+    int set_pin(const char* pin, char *resp_buf, int size_buf);
+    int enable_SSL(const char *filename, char *resp_buf, int size_buf);
```
[the remark this answers is not in the document]
Comment (azeemshatp) [reply src/SIM800-AT.h]: Agreed and modified

**Pull request merged**